### ISSUE TITLE
Implement SLURM support

### ISSUE THREAD

**Issue by samuell** [scipipe/scipipe#38]

**Comment by valentin-krasontovitsch** [comment]
Considering moving away from ruffus written in python to a golang pipeline framework, and slurm is what we use for batching... is there any prognosis on this? a roadmap? what would have to be done - are you accepting merge requests?

**Comment by valentin-krasontovitsch** [comment]
Wait, it's already perfectly possible to use slurm just by using a go slurm library and writing go code directly in the workflow, correct? What is this issue about, precisely? I'm confused ^^

**Comment by samuell** [comment]
Hi @valentin-krasontovitsch , thanks for checking out scipipe.

What we are doing right now, while pondering the best way to add proper SLURM support, is to just use the prepend feature, and add the SLURM `salloc` command ther (as shown in http://scipipe.org/howtos/hpc/). It actually works very robustly, but it would be sure great to have something a tad bit more integrated.

I will check the go SLURM library (wasn't aware of it yet :) ), but I'm also leaning towards keeping everything as simple as possible, fighting a bit to avoid dependencies etc ... as I think this will help make scipipe easier to maintain in the long run, with limited academic funding :P

**Comment by samuell** [comment]
Ah, but yes, you should be able to call Go/SLURM library code from within your [custom Go-functions](http://scipipe.org/howtos/golang_components/), for sure, yes!

**Comment by samuell** [comment]
Otherwise, I'm right now looking hard into introducing a bit more of modularity / pluggability in the code, so that it would be easier for anyone to add custom "runners", according to some interface, without needing to rewrite the core. Hope to get this fixed before 1.0 in a few months.

**Comment by valentin-krasontovitsch** [comment]
I'm not sure about the slurm library myself, have only looked at it shortly, but it seems like a mature project (just judging by the fact that they seem to have a version 1 and a version 2 ^^ ) - always hard to weigh wether you want to introduce a dependency and maintain that, or write it yourself and maintain that... 

but I'm just wondering what kind of features your thinking of that would make slurm more "integrated"? Or more abstractly, any way of running commands / any runner? have you already compiled a (mental) list?

cause at the moment, as you said as well, everybody's just free to use whatever library they want in custom go functions, right? what is that lacking in "integration" compared to the "normal" way of running a command?

**Comment by samuell** [comment]
@valentin-krasontovitsch 

> I'm not sure about the slurm library myself, have only looked at it shortly, but it seems like a mature project (just judging by the fact that they seem to have a version 1 and a version 2 ^^ ) - always hard to weigh wether you want to introduce a dependency and maintain that, or write it yourself and maintain that...

Indeed.

> but I'm just wondering what kind of features your thinking of that would make slurm more "integrated"? Or more abstractly, any way of running commands / any runner? have you already compiled a (mental) list?

Having data stored in a bit more structured way would help with things like structured logging. 

This info could then also in theory be mined to help provide some smarter way to provide good "guesstimates" for the time limits to put on the SLURM job ... something that can be quite a manual hassle otherwise, where running times can depend a lot on the size of the input data etc. 

> cause at the moment, as you said as well, everybody's just free to use whatever library they want in custom go functions, right? what is that lacking in "integration" compared to the "normal" way of running a command?

That's a very good point though! The custom Go function feature might be exactly the right place allow custom runners. Perhaps just putting the runner in a struct with a few metadata fields would be enough to provide a good generic solution.

**Comment by samuell** [comment]
Just wanted to report that I think I have now seen the light regarding how to implement the required pluggability. It goes quite along the lines of your way of just using the customExexute method, @valentin-krasontovitsch , but I realized that we can make this in an even more documented and structured way, by allowing to plug in your custom "Executor".

This "Executor" will just be a Go interface of type:

```go
type Executor interface {
    Execute(t *scipipe.Task)
}
```

... possibly with any other minor methods.

Then, one could have just create different executors, using structs, with their respective factory methods ... so one could just go, in the workflow:

```go
foo := wf.NewProc("foo", "echo foo > {o:out}")
foo.Executor = NewSlurmExecutor("b2013222", "core", "2", "1:00:00", "MyFooJob")
```

... which would be implemented along the lines of (mostly notes for myself):

```go
type SlurmExecutor struct {
    account string
    partition string
    numCores int
    runTime string
    jobName string
}

func (e *SlurmExecutor) Execute(t *scipipe.Task) {
    slurmCmd := fmt.Sprintf("salloc -A %s -p %s -n %s -t %s -J %s", e.account, e.partition, e.numCores, e.runTime, e.jobName)
    command := slurmCmd + " " + t.Command
    // Execute command
}

func NewSlurmExecutor(account string, partition string ...) *SlurmExecutor {
    executor := &SlurmExecutor{
        account: account,
        // ... etc etc ...
    }
    return executor
}
```

**Comment by valentin-krasontovitsch** [comment]
Sorry, I got a bit side-tracked with other projects...

That sounds awesome, and I like it a lot ^^

Thanks for following up!

I hope I will actually be able to get to this and use it one day...

**Comment by cvlvxi** [comment]
Hey both.

Just wondering if there was more traction made with this? I'm interested in running on a Torque cluster instead of Slurm. Any advise would be great!

**Comment by valentin-krasontovitsch** [comment]
I'm sorry, but we're currently trying to switch to a WDL (workflow definition language) based pipeline (so we can just use what the team at the Broad Institute - GATK is churning out), with Cromwell, and Kubernetes (via TESK) as a backend.

But if all that fails, we might reconsider scipipe!

**Comment by samuell** [comment]
> Hey both.
> 
> Just wondering if there was more traction made with this? I'm interested in running on a Torque cluster instead of Slurm. Any advise would be great!

Hi @cvlvxi ! 

Do you know if Torque has a command that blocks while the job is queuing and running, just like the [salloc](https://slurm.schedmd.com/salloc.html) command does with SLURM?

If so, it should be trivial to prepend the commands with this command, in place of salloc, along the lines of this page in the docs: [SciPipe Docs / Interacting with an HPC resource manager](https://scipipe.org/howtos/hpc/)

**Comment by cvlvxi** [comment]
Yep. The equivalent to salloc is qsub. Thanks I'll give it a go :)

Cheers!

**Comment by samuell** [comment]
> Yep. The equivalent to salloc is qsub. Thanks I'll give it a go :)

Great - Hope it works! :)

Cheers
